sys/linux/bpf.txt

```diff
@@ -383,16 +397,19 @@ bpf_insn_generic {
 	imm	int32
 }
 
-bpf_insn_ldst {
-	code_class	flags[bpf_ldst_insn, int8:3]
-	code_size	flags[bpf_ldst_size, int8:2]
-	code_mode	flags[bpf_ldst_mode, int8:3]
-	dst		flags[bpf_reg, int8:4]
-	src		flags[bpf_reg, int8:4]
-	off		flags[bpf_insn_offsets, int16]
-	imm		flags[bpf_insn_immediates, int32]
+type bpf_insn_ldst_t[CLASS, SZ, MODE, DST, SRC, OFF, IMM] {
+	code_class	CLASS
```

Comment:
Only the value should change, but if we keep the bitfield in the template itself, then we will be forced to define a new flag every time we want to pass a specific class (which is likely to be very common).

So instead of having `const[BPF_STX, int8:3]`, we'll have to declare a new flag `bpf_stx = BPF_STX` for every new value. Is that really worth it?

Reply:
Ah, right. That's unfortunate.

What do you think about making class field type a template argument?

```
type bpf_insn_ldst_t[CLASS_TYPE, CLASS, SZ, MODE, DST, SRC, OFF, IMM] {
       code_class CLASS_TYPE[CLASS, int8:3]
```

Reply:
I think that's a good middle ground, but we'll have to do that for `SZ` and `MODE` too :disappointed:
```
type bpf_insn_ldst_t[CLASS_TYPE, CLASS, SZ_TYPE, SZ, MODE_TYPE, MODE, DST, SRC, OFF, IMM]
```

It feels like an ideal solution would be to have some sort of `flag(X)` construct that makes a flag out of `X` if it isn't one already.

Reply:
Oh, interesting, thanks for the details. Then it won't solve all problems indeed :(

Though even with `type bpf_insn_ldst_t[CLASS_TYPE, CLASS, SZ_TYPE, SZ, MODE_TYPE, MODE, DST, SRC, OFF, IMM]` type instantiations already look somewhat better:

```
type bpf_insn_ldst bpf_insn_ldst_t[flags[bpf_ldst_insn, int8:3], flags[bpf_ldst_size, int8:2],
flags[bpf_ldst_mode, int8:3], flags[bpf_reg, int8:4], flags[bpf_reg, int8:4],
flags[bpf_insn_offsets, int16], flags[bpf_insn_immediates, int32]]
```

```
type bpf_insn_ldst bpf_insn_ldst_t[flags, bpf_ldst_insn, flags, bpf_ldst_size, flags,
bpf_ldst_mode, bpf_reg, bpf_reg, bpf_insn_offsets, bpf_insn_immediates]
```

Otherwise I think it all just beats the purpose of templates. If we write a comma-separated list of N full field descriptions here, then it's just easier to define several separate structures from scratch. It will all be much more readable that way.

>  It feels like an ideal solution would be to have some sort of flag(X) construct that makes a flag out of X if it isn't one already.

That sounds interesting indeed. We could also let `flags[]` type directly accept a single const instead of a list, e.g. `flags[BPF_STX, int8:3]`.

I don't know if it really matters here, but technically `const` and `flags` have somewhat different semantics in `syzkaller`: `flags` are mutated (they are also sometimes set to completely random int values), `const`s **always** have the same value.

These infrequent mutations might not really be a big problem per se, but in this PR you'd probably want syzkaller to put insane value to fields other than class, type and mode (?).

But still `flags[single-const]` support is so much easier to do that the concern above is probably just not worth worrying about.

Reply:
> Though even with `type bpf_insn_ldst_t[CLASS_TYPE, CLASS, SZ_TYPE, SZ, MODE_TYPE, MODE, DST, SRC, OFF, IMM]` type instantiations already look somewhat better:

Agree. I've pushed the changes.

> We could also let `flags[]` type directly accept a single const instead of a list, e.g. `flags[BPF_STX, int8:3]`.

I thought about that approach as well, but I'm a bit scared it might make it easier to introduce errors. It's already fairly easy to introduce errors in the description; except if we have tests, we're unlikely to notice. Having `flags[]` accept a `const` in addition to a list feels a bit like we're making the "type system" looser. Yet the type checks have been the main way syzkaller surfaced bugs in my changes so far. Hence my proposal for some sort of explicit cast.

Reply:
> I thought about that approach as well, but I'm a bit scared it might make it easier to introduce errors. 

Good point. I've filed https://github.com/google/syzkaller/issues/4269